src/opencl/utils.rs

```diff
@@ -1,21 +1,18 @@
+use std::convert::{TryFrom, TryInto};
 use std::fmt::Write;
 
 use lazy_static::lazy_static;
 use log::{debug, warn};
 use opencl3::device::DeviceInfo::CL_DEVICE_GLOBAL_MEM_SIZE;
 use sha2::{Digest, Sha256};
 
-use super::{Brand, Device, GPUError, GPUResult};
-
-const AMD_DEVICE_VENDOR_STRING: &str = "AMD";
-const NVIDIA_DEVICE_VENDOR_STRING: &str = "NVIDIA Corporation";
+use super::{Device, GPUError, GPUResult, Vendor};
 
 fn get_bus_id(d: &opencl3::device::Device) -> Result<u32, GPUError> {
-    let vendor = d.vendor()?;
-    match vendor.as_str() {
-        AMD_DEVICE_VENDOR_STRING => d.pci_bus_id_amd().map_err(Into::into),
-        NVIDIA_DEVICE_VENDOR_STRING => d.pci_bus_id_nv().map_err(Into::into),
-        _ => Err(GPUError::DeviceBusId(vendor)),
+    let vendor = Vendor::try_from(d.vendor()?.as_str())?;
+    match vendor {
+        Vendor::Amd => d.pci_bus_id_amd().map_err(Into::into),
```

Comment:
why not use `?`

Reply:
Because we return a `Result`, this way I don't need to wrap it again in an `Ok`.